File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/297. Serialize and Deserialize Binary Tree.cpp
// 297. Serialize and Deserialize Binary Tree
// https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
#include <iostream>
#include <string>
#include <stack>
#include <sstream>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "null";
        return to_string(root->val) + "," + serialize(root->left) + "," + serialize(root->right);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return deserializeHelper(data, pos);
    }
    
private:
    TreeNode* deserializeHelper(const string &data, int &pos) {
        int next_comma = data.find(',', pos);
        string value = data.substr(pos, next_comma - pos);
        pos = next_comma + 1;
        if (value == "null") return nullptr;
        TreeNode *node = new TreeNode(stoi(value));
        node->left = deserializeHelper(data, pos);
        node->right = deserializeHelper(data, pos);
        return node;
    }
};

int main() {
    Codec codec;
    string data = "1,2,3,null,null,4,5,null,null,null,null";
    TreeNode *root = codec.deserialize(data);
    cout << codec.serialize(root) << endl;
    return 0;
}

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/297. Serialize and Deserialize Binary Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string."""
        def dfs(node):
            if not node:
                return "None,"
            return str(node.val) + "," + dfs(node.left) + dfs(node.right)
        return dfs(root)

    def deserialize(self, data):
        """Decodes your encoded data to tree."""
        def dfs(nodes):
            val = next(nodes)
            if val == "None":
                return None
            node = TreeNode(int(val))
            node.left = dfs(nodes)
            node.right = dfs(nodes)
            return node
        return dfs(iter(data.split(',')))

# 测试用例
codec = Codec()
tree = TreeNode(1, TreeNode(2), TreeNode(3, TreeNode(4), TreeNode(5)))
data = codec.serialize(tree)
print(data)  # 输出序列化的字符串
deserialized_tree = codec.deserialize(data)
print(codec.serialize(deserialized_tree))  # 验证反序列化是否正确
